### ISSUE TITLE
MSTIS analysis, error when using collective variable cache_all

### ISSUE THREAD

**Issue by Asagodi** [choderalab/openpathsampling#373]
When trying to use cache_all to speed up the analysis, I get an error for the Collective variable caching. Should storage be added as an attribute when calling the function (as there is no such thing in de function definition) or should that be added to the function definition (the description state that a storage should be possible to add).

The other part of the file, is like the mstis.ipynb in examples. Here is the line causing the error:
map(lambda x : x.cache_all(storage), storage.cvs)
The error:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-43-76b50f3a31d4> in <module>()
      4 storage.pathmovechanges.cache_all()
      5 storage.steps.cache_all()
----> 6 map(lambda x : x.cache_all(storage), storage.cvs)

<ipython-input-43-76b50f3a31d4> in <lambda>(x)
      4 storage.pathmovechanges.cache_all()
      5 storage.steps.cache_all()
----> 6 map(lambda x : x.cache_all(storage), storage.cvs)

TypeError: cache_all() takes exactly 1 argument (2 given)

**Comment by dwhswenson** [comment]
Assigning this to @jhprinz, since storage is his thing. However:

1. How recent is your version of OPS? Storage had major changes in #300 (merged on Nov 3).
2. If you're using a version after #300, you should be using `AnalysisStorage` to load files for analysis. This will automatically handle all the caching, so you shouldn't need these lines.

**Comment by Asagodi** [comment]
I recently updated OPS and have the version with the changes. The runs are really fast now! AnalysisStorage indeed works, however, the calculation of the rates is a lot slower now. 

Further, I would like to access timings in storage.step.change.details. How can I do this easily?

**Comment by jhprinz** [comment]
What is slow for rate computation. It should be about a few seconds to half a minute if you have really lots of trajectories. That is how long it takes on my computer usually.

If it is really slow like several minutes it could be that you do not store the cached CVs properly in the file.

Could you send me the notebook(s) somehow so that I can have a look?

**Comment by dwhswenson** [comment]
The rate calculation requires determination of the flux from the minus moves. All the minus trajectories must be loaded and checked against several volumes. This is the main cost in the rate.

If there is a faster way to load all the snapshots from a trajectory, that would speed this up. If it is slow, I assume the reason is that right now each snapshot individually goes through the process of a cache miss and load from disk; if that's the case, batching this into a per-trajectory load from disk should speed things up.

The place to do it would be [`tis_analysis.py:TISTransition.minus_move_flux`, after the line `for step in minus_steps:`](https://github.com/choderalab/openpathsampling/blob/9a61257c89c0bbe8273f4569afd1f07a990ebc2c/openpathsampling/analysis/tis_analysis.py#L479) (link pointing to the most recent commit to master, where this is line 479). @jhprinz : is there a quick and easy way to make this faster?

**Comment by Asagodi** [comment]
It is about 30 minutes. It was not possible to attach the notebook in the post, I will send it next week.

**Comment by dwhswenson** [comment]
Assigning this ASAP over the speed concerns. 

@Asagodi : one of the easier ways to share notebooks is through GitHub's gists: start a gist at http://gist.github.com, and then you can post the link here. It even supports drag-and-drop uploads. Unfortunately, file size limitations will probably prevent you from posting large `.nc` files. Those are best uploaded to DropBox, Google Drive, or similar and sharing a link.

**Comment by Asagodi** [comment]
The script: https://gist.github.com/Asagodi/a73473325a51e69bf705#file-4d_04-ipynb

**Comment by jhprinz** [comment]
Thanks. This looks really cool. Have never seen such a big simulation. 

Comment on the speed concerns.
Alright. The problem is potentially that you have too many snapshots. :) I set the analysis cache to use only 100k snapshots and you have 1.5M. This will work (as you realize) but be slow!!!. 
I assume that it could help to increate the snapshot cache, but what you actually want is to increase the cache for the collective variables.

There is no easy way todo that yet, but I will make a PR to fix that. If you have an installation where you can edit the code go to `chaindict.py` and look close for the class `StoredDict`. This takes care of caching cvs in a storage.
```
class StoredDict(ChainDict):
    """
    ChainDict that has a store attached and return existing values from the store
    """
    def __init__(self, key_store, value_store, main_cache, cache=None):
        """
        Parameters
        ----------
        key_store : storage.Store
            the store that references usable keys
        value_store : storage.Variable
            the store that references the store variable to store the values by index
        cache : dict-like or cache, default: None
            the cache used to access stored values faster. If None an LRUCache with
            100000 entries is used
        """
        super(StoredDict, self).__init__()
        self.value_store = value_store
        self.key_store = key_store
        self.main_cache = main_cache
        self.max_save_buffer_size = None
        if cache is None:
            cache = LRUCache(1000000) # increase this number to allow more values to be cached!
        self.cache = cache
        self.storable = set()
        self._last_n_objects = 0

```

Hope this helps. So, put simply, you are using too much data for the defaults.

**Comment by Asagodi** [comment]
Thank you! 

What surprised me was that the calculations were slower after the upgrading to the new version of OPS. 
Just out of curiosity: Are the rate calculated in a very different way now?

**Comment by jhprinz** [comment]
Yes, the old version did not have a limited cache, it was unlimited and so it should have been faster.

**Comment by jhprinz** [comment]
I am currently working on a better storage for the ToyDynamics which made the simulations also faster. I hope this is ready soon. It works by making the Snapshot objects simpler and it felt like being much faster. More to come soon!

**Comment by dwhswenson** [comment]
No, there's no difference in the rate calculation. Only in the way things or stored. Previously, *everything* was always in RAM. For a small system like yours, where each snapshot is only a few bytes, that's no problem. But for large enough systems, you can easily fill RAM with only a few hundred trajectories. Not removing snapshots from memory would have made the whole program crash.

The caching tricks @jhprinz developed avoid that problem, but at the cost of making things a little trickier for the kinds of calculation you're doing.

**Comment by jhprinz** [comment]
I will add the possibility to set the way CVs are cached manually. Say you have lots of RAM you can chose to go RAM only as before. In principle this was a good test that it still works. In my tests where I used a very small cache it was also very slow.

**Comment by dwhswenson** [comment]
@jhprinz : I'm a little unsure whether LRU is the best approach, as I think about it. Consider what's happening as you calculate the minus move: you must load every snapshot from a minus trajectory. Once the cache is full, do you just add/remove one at a time? 

It seems that a temporary cache load of all frames, followed by a cache clear, will be better for this kind of analysis. If the frames over-fill the cache, the best thing to do is to fill as many as possible, use them, then empty and re-fill the cache with as many as possible. If you're managing the cache at near-full capacity, that will cost a lot more effort for very little value.

**Comment by jhprinz** [comment]
My question would be what these configurations/snapshots are used for. If they are used with CVs and these CVs are already cached and stored, then you do not need to load any snapshots, this is handled completely without loading. This is why in normal analysis caching we do not load all snapshots. Only trajectories.

If however you still need to compute these CVs then best is to loop over all snapshots using a CV to cache the CVs as fast as possible. After that it should be fast. Problem with Abel was that the CV cache became slow since it was too small and then, one by one is loaded and the cache is updated. This takes much time and is not optimal.

Last answer on how to get the minus trajectories, this would be

```
minus_trajs = [sample.trajectory for sample in storage.samples if sample.ensemble is my_minus_ensemble]
for traj in minus_trajs:
    for snapshot in traj:
        # this is a real snapshot not lazy and hence slow!

for traj in minus_trajs:
    for snapshot in traj.as_proxies():
        # this are lazy snapshots these can be used in CVs, ensembles, etc as well, 
        # but will only be loaded if necessary. Should be much faster in case you have
        # lots of precomputed CVs
```

**Comment by jhprinz** [comment]
> How does that work if the objects you want to chunk-load are not contiguous in storage?

The same way this worked before: Load all chunks so that you cover what you need. You need to choose a good chunksize though. Usually you have at least a certain amount of contiguous frames. After each loaded chunk pick the objects you need from it and load the next chunk. Finally return the list. Difference to now is that we currently use chunk size one always.

**Comment by dwhswenson** [comment]
> My question would be what these configurations/snapshots are used for.

Hmm... yes, I guess the CV associated with the volume should be already calculated. Good point. The volume itself, however, is not going to be stored as a CV. So yes, you're right: the CV should have been precalculated, and it was just loading the CVs (and possibly the snapshots associated).

That form of chunking still seems a bit inefficient: sometimes we gather unnecessary data. I was trying to figure out whether there was some iterator-based way to get only what you need.

**Comment by jhprinz** [comment]
What do you mean by unnecessary data? If you want all stored trajectories that are sampled of ensemble `xyz` use what I wrote. If you only want all resulting samples from steps you would use
```
minus_traj = [sample for step in storage.steps for sample in step.active if snapshot.samples is xyz]
```

if you want all trajectories only once use
```
set(minus_traj)
```
it should be really that easy and not much more difficult. If you have several exclusive ensembles then it can be improved. Since you know when you find one ensemble you can exclude the others, but that only makes sense if you use `sample in xyz` and not `sample.ensemble is xyz` because the second is pracitually instantaneous. 
```
ensemble_list = [minus_1, minus_2, ...]
ensemble_per_sample = dict()
samples_in_ensemble = {ensemble:[] for ensemble in ensemble_list}
unique_samples = list(set(sample for step in storage.steps for sample in step.active))
for sample in unique_samples:
    for ensemble in ensemble_list:
        if ensemble(sample):
            ensemble_per_sample[sample] = ensemble #pick which you
            samples_in_ensemble[ensemble].append(sample) #prefer

# If you want non unique you need to loop over all the samples a second time using your results
```

**Comment by dwhswenson** [comment]
> What do you mean by unnecessary data?

My understanding of your suggestion was that, when you tried to load something (e.g. a snapshot) that wasn't in the cache, you would load a chunk of data into the cache (e.g., many consecutively stored snapshots).

If you only need one snapshot, or if the snapshots you need are not stored anywhere near consecutively, then you spend significant time loading snapshots that you never use into the cache. That's what I mean by unnecessary data. 

It might also useful for the chunking behavior to be changeable. We wouldn't expect the average user to do this, but the developer of an analysis routine might take advantage of this to optimize for a specific analysis. For example, although loading objects in sequence benefits from chunking, random selection might get slowed down by chunking.

**Comment by jhprinz** [comment]
I see. Yes, you are right of course. My feeling is that we mostly load stuff that is close together. Like in trajectories you will have mostly chunks that could be from the same engine run. If you have lots of random access I would prefer caching as much as possible. We could also set the chunk size manually for special cases.

**Comment by Asagodi** [comment]
@jhprinz : The description of the StoredDict class states that default cache is 100000 (10⁵), while the implemented function sets it to 1000000 (10⁶). 
Which is the ideal number?

**Comment by jhprinz** [comment]
Okay, that was a mistake. I think I used one million everywhere and Abel made more than a million. The stored dict is basically a list since it maps integers to CV values. I thought to increase the number even.

**Comment by Asagodi** [comment]
Changing cache helped indeed.
Allowing to cache all data at once overfilled RAM (~100% + 60% Swap memory), but helped to succeed the rate calculations. With a ~3000000 snapshot dataset, a cache of 1000000 failed on my computer, while a cache of 5000000 succeeded (in 3.5 hours).

**Comment by jhprinz** [comment]
Okay, that is pretty cool... I think I never thought there would be that many snapshots. But good to know that it works. Out of curiosity, how many CVs did you use?

I think we can still improve on the speed, but that involves a deeper insight into storage. I think there are even faster ways to get all the data needed for rate computations. As long as you have already computed all the CVs before you start analyzing. Otherwise this will be really slow.

**Comment by dwhswenson** [comment]
@jhprinz : I've run simulations that probably had ~10 million trajectories, averaging ~100 frames per trajectory (all with 2-way shooting, so all those snapshots were new), and 10 different order parameters. So that's ~1 billion frames, each consisting of 25 atoms in 2D (so 50 floats per frame). And I needed data from 6 such simulations for the paper (and probably ran a couple dozen over the course of getting the data I needed).

This is why I've said it would be a good idea to have the option to ***not*** save all the trajectory data. Toy systems can have more snapshots than you want to store.

**Comment by Asagodi** [comment]
I used 4 different CV volumes and 4 CV functions for the 4 states, and further 4 CV volume sets for for the interfaces with 8 intervals.

**Comment by jhprinz** [comment]
Just catching up. What is the status on this? Can we close this? Or do I need to add something for better caching, load/save?

**Comment by dwhswenson** [comment]
The main things that I think we need (and I think we already have) are:

1. A way to change the number of items allowed in the cache
2. A way to change the chunking size (if we have chunking implemented)

This is enough to allow a user to adjust those settings to give optimal speed. Since "optimal" depends on the nature of the system being studied as well as the hardware being used, the best we can do is give reasonable defaults.

All of that is already implemented, right? I mean, we might play with the defaults, but I thought those basic capabilities were already in place. If so, then as far as I'm concerned, we can close this one. And it looks like, from @Asagodi's comment on Nov 27, his problem is also solved.

(Adding the ability to only save full trajectories every Nth MC step is, I think, a good idea -- but for the future. It would require some significant changes.)

**Comment by Asagodi** [comment]
@jhprinz The rate calculations are fine. The simulation speed seems even to have increased!

**Comment by jhprinz** [comment]
> 1. A way to change the number of items allowed in the cache
> 2. A way to change the chunking size (if we have chunking implemented)

1. Yes
2. No, but it is on my list. Not a priority right now. It should improve read/write though. I would call this a OPS 1.5 feature.

**Comment by dwhswenson** [comment]
On point 2 (chunking): "no" meaning chunking isn't implemented, or that the value can't be modified by the user?

**Comment by jhprinz** [comment]
It is not implemented. I worked out most of the details, but never got to really implement it. netcdf does chunking natively so if you set chunking wrong it will always load lots of unnecessary data.

You might remember the 100MB file without any stored data :)

**Comment by dwhswenson** [comment]
> You might remember the 100MB file without any stored data :)

Oh yes, that old thing. :-) Glad we fixed that, at least!

That said, I'll close this. (If it had been "chunking implemented but not user-changeable", I would have closed this and opened a new issue as a reminder of that -- as it stands, I assume you plan to make chunk sizes user-changeable when you implement it!).
